src/pykeen/datasets/analysis.py

```diff
@@ -1,72 +1,103 @@
-    'iter_binary_patterns',
-    'iter_ternary_patterns',
-    'triple_set_hash',
+    "relation_count_dataframe",
+    "entity_count_dataframe",
+    "entity_relation_co_occurrence_dataframe",
+    "calculate_relation_functionality",
+    # relation typing
+    "relation_pattern_types",
+    "relation_cardinality_types"
 ]
 
-SUBSET_LABELS = ('testing', 'training', 'validation', 'total')
+# constants
+COUNT_COLUMN_NAME = "count"
+ENTITY_ID_COLUMN_NAME = "entity_id"
+ENTITY_LABEL_COLUMN_NAME = "entity_label"
+RELATION_ID_COLUMN_NAME = "relation_id"
+RELATION_LABEL_COLUMN_NAME = "relation_label"
+SUBSET_COLUMN_NAME = "subset"
 
+#: fixme: deprecated
+SUBSET_LABELS = ("testing", "training", "validation", "total")
 
-# PatternMatch = namedtuple('PatternMatch', ['relation_id', 'pattern_type', 'support', 'confidence'])
 
-class PatternMatch(NamedTuple):
-    """A pattern match tuple of relation_id, pattern_type, support, and confidence."""
+def _get_mapped_triples(dataset: Dataset, parts: Collection[str]) -> Collection[Tuple[int, int, int]]:
+    return torch.cat([
+        dataset.factory_dict[part].mapped_triples
+        for part in parts
+    ], dim=0).tolist()
 
-    relation_id: int
-    pattern_type: str
-    support: int
-    confidence: float
 
+def _normalize_parts(dataset: Dataset, parts: Union[None, str, Collection[str]]) -> Collection[str]:
+    if parts is None:
+        parts = dataset.factory_dict.keys()
+    elif isinstance(parts, str):
+        parts = [parts]
+    # unique
+    return list(set(parts))
 
-def get_id_counts(id_tensor: torch.LongTensor, num_ids: int) -> numpy.ndarray:
-    """Create a dense tensor of ID counts.
 
-    :param id_tensor:
-        The tensor of IDs.
-    :param num_ids:
-        The number of IDs.
+def _add_labels(
+    label_to_id: Optional[Mapping[str, int]],
+    df: pd.DataFrame,
+    id_column: str,
+    label_column: str,
+) -> pd.DataFrame:
+    if label_to_id is None:
+        return df
+    return pd.merge(
+        left=df,
+        right=pd.DataFrame(
+            data=list(label_to_id.items()),
+            columns=[label_column, id_column],
+        ),
+        on=id_column,
+    )
 
-    :return: shape: (num_ids,)
-         The counts for each individual ID from {0, 1, ..., num_ids-1}.
-    """
-    unique, counts = id_tensor.unique(return_counts=True)
-    total_counts = numpy.zeros(shape=(num_ids,), dtype=numpy.int64)
-    total_counts[unique.numpy()] = counts.numpy()
-    return total_counts
 
+_add_entity_labels = functools.partial(
+    _add_labels,
+    id_column=ENTITY_ID_COLUMN_NAME,
+    label_column=ENTITY_LABEL_COLUMN_NAME,
+)
 
-def relation_count_dataframe(dataset: Dataset) -> pd.DataFrame:
-    """Create a dataframe with relation counts for all subsets, and the full dataset.
+_add_relation_labels = functools.partial(
+    _add_labels,
+    id_column=RELATION_ID_COLUMN_NAME,
+    label_column=RELATION_LABEL_COLUMN_NAME,
+)
+
+
+def relation_count_dataframe(
```

Comment:
@cthoyt this PR is beginning to become big again :smile: 

What is your comment on this kind of flexibility: dataset (+parts?), triples factory, mapped_triples, ...

Reply:
tbh I don't see why someone would ever do this analysis on a dataset and not use all 3 triples factories in it

Reply:
One reason would be to have only some triples, and not a pre-split dataset. Then, passing a single TriplesFactory / only ID-based triples may be convenient.

Reply:
I thought we already split the implementation up so the triples interface is available through pykeen.triples.analysis

Reply:
I went back to having one method which takes all of the arguments, and calls itself with the reduced input